### PR TITLE
ANet datatools

### PR DESCRIPTION
**Pull request by kennymckormick**

### CODE REVIEW

File: configs/recognition/tsn/tsn_r50_320p_1x1x8_150e_activitynet_clip_flow.py [hunk 1]
```diff
@@ -80,6 +80,7 @@
         filename_tmpl='flow_{}_{:05d}.jpg',
         with_offset=True,
         modality='Flow',
+        start_index=0,
```
Comment (innerlee) [inline configs/recognition/tsn/tsn_r50_320p_1x1x8_150e_activitynet_clip_flow.py:83]: i thought we moved `start_index` to dataset? @dreamerlin

Comment (dreamerlin) [reply configs/recognition/tsn/tsn_r50_320p_1x1x8_150e_activitynet_clip_flow.py:83]: Yes. Move it to [here](https://github.com/open-mmlab/mmaction2/blob/master/mmaction/datasets/base.py#L51) and the [old one](https://github.com/open-mmlab/mmaction2/blob/master/mmaction/datasets/pipelines/loading.py#L62-L65) has been deprecated

File: tools/data/activitynet/action_name.csv [hunk 2]
```diff
@@ -0,0 +1,201 @@
+action
```
Comment (dreamerlin) [inline tools/data/activitynet/action_name.csv:1]: label_map.txt

File: tools/data/activitynet/download.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
+# This scripts is copied from
+# https://github.com/activitynet/ActivityNet/blob/master/Crawler/Kinetics/download.py  # noqa: E501
+import os
+import subprocess
+
+import mmcv
+
+import ssl  # isort:skip
+
+from joblib import Parallel, delayed  # isort:skip
```
Comment (dreamerlin) [inline tools/data/activitynet/download.py:10]: Why use `isort:skip`?

File: tools/data/activitynet/download.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        File path where the video will be stored.
+    """
+    # Defensive argument checking.
+    assert isinstance(video_identifier, str), 'video_identifier must be string'
+    assert isinstance(output_filename, str), 'output_filename must be string'
+    assert len(video_identifier) == 11, 'video_identifier must have length 11'
+
+    status = False
+
+    if not os.path.exists(output_filename):
```
Comment (dreamerlin) [inline tools/data/activitynet/download.py:38]: `os.path` -> `osp`

File: tools/data/activitynet/generate_rawframes_filelist.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
+import os
```
Comment (dreamerlin) [inline tools/data/activitynet/generate_rawframes_filelist.py:1]: Where is this file used?

Comment (kennymckormick) [reply tools/data/activitynet/generate_rawframes_filelist.py:1]: to generate anet_train_video.txt, ... which are used for anet finetuning

File: configs/recognition/tsn/tsn_r50_320p_1x1x8_150e_activitynet_clip_flow.py (hunk 1 above)
Comment (kennymckormick) [reply configs/recognition/tsn/tsn_r50_320p_1x1x8_150e_activitynet_clip_flow.py:83]: a little confused, I think I have written start_index in dataset config.

Comment (innerlee) [reply configs/recognition/tsn/tsn_r50_320p_1x1x8_150e_activitynet_clip_flow.py:83]: oh right my mistake

File: tools/data/activitynet/download.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
+# This scripts is copied from
+# https://github.com/activitynet/ActivityNet/blob/master/Crawler/Kinetics/download.py  # noqa: E501
```
Comment (innerlee) [inline tools/data/activitynet/download.py:2]: mention its licence

File: tools/data/activitynet/download.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+def main(input_csv, output_dir, anno_file, num_jobs=24):
+    # Reading and parsing ActivityNet.
+    youtube_ids = parse_activitynet_annotations(input_csv)
+
+    # Creates folders where videos will be saved later.
+    if not os.path.exists(output_dir):
+        os.makedirs(output_dir)
+    # Download all clips.
+    if num_jobs == 1:
+        status_lst = []
```
Comment (dreamerlin) [inline tools/data/activitynet/download.py:105]: `lst` -> `list`

File: tools/data/activitynet/generate_rawframes_filelist.py [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+    anet_annotations = mmcv.load(anno_file)
+
+    videos = open(video_list).readlines()
+    videos = [x.strip().split(',') for x in videos]
+    attr_names = videos[0][1:]
+    # the first line is 'video,numFrame,seconds,fps,rfps,subset,featureFrame'
+    attr_names = [x.lower() for x in attr_names]
+    attr_types = [int, float, float, float, str, int]
+
+    vidannos = {}
```
Comment (dreamerlin) [inline tools/data/activitynet/generate_rawframes_filelist.py:23]: `vidannos` -> `video_annos`

File: tools/data/activitynet/generate_rawframes_filelist.py [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+        data = {}
+        for attr_name, attr_type, attr_val in zip(attr_names, attr_types,
+                                                  line[1:]):
+            data[attr_name] = attr_type(attr_val)
+        vidannos[name] = data
+
+    # only keep downloaded videos
+    vidannos = {k: v for k, v in vidannos.items() if k in anet_annotations}
+    # update numframe
+    for vid in vidannos.keys():
```
Comment (dreamerlin) [inline tools/data/activitynet/generate_rawframes_filelist.py:35]: `for video in video_annos:`

File: tools/data/activitynet/generate_rawframes_filelist.py [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+        pth = osp.join(rawframe_dir, vid)
+        num_imgs = len(os.listdir(pth))
+        # one more rgb img than flow
+        assert (num_imgs - 1) % 3 == 0
+        num_frames = (num_imgs - 1) // 3
+        vidannos[vid]['numframe'] = num_frames
+
+    anet_labels = open('action_name.csv')
+    anet_labels = [x.strip() for x in anet_labels]
```
Comment (dreamerlin) [inline tools/data/activitynet/generate_rawframes_filelist.py:44]:
```
with open('action_name.csv', 'r') as f:
    anet_labels = [x.strip() for x in f]
```

File: tools/data/activitynet/generate_rawframes_filelist.py [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+
+    train_vids = {
+        k: vid
+        for k, vid in vidannos.items() if vid['subset'] == 'training'
+    }
+    val_vids = {
+        k: vid
+        for k, vid in vidannos.items() if vid['subset'] == 'validation'
+    }
```
Comment (dreamerlin) [inline tools/data/activitynet/generate_rawframes_filelist.py:53]:
```
train_videos, val_videos = {}, {}
for k, video in video_annos.items():
    if video['subset'] == 'training':
        train_videos[k] = video
    elif video['subset'] == 'validation':
        val_videos[k] = video
```

File: tools/data/activitynet/generate_rawframes_filelist.py [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+
+    def simple_label(vidix):
```
Comment (dreamerlin) [inline tools/data/activitynet/generate_rawframes_filelist.py:55]: `vidix` -> `video_idx`

File: tools/data/activitynet/generate_rawframes_filelist.py [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+        anno = anet_annotations[vidix]
+        lb = anno['annotations'][0]['label']
```
Comment (dreamerlin) [inline tools/data/activitynet/generate_rawframes_filelist.py:57]: `lb` -> `label`

File: tools/data/activitynet/generate_rawframes_filelist.py [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+        return anet_labels.index(lb)
+
+    train_lines = [
+        k + ' ' + str(train_vids[k]['numframe']) + ' ' + str(simple_label(k))
+        for k in train_vids.keys()
```
Comment (dreamerlin) [inline tools/data/activitynet/generate_rawframes_filelist.py:62]: `for k in train_vids`

File: tools/data/activitynet/generate_rawframes_filelist.py [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+    ]
+    val_lines = [
+        k + ' ' + str(val_vids[k]['numframe']) + ' ' + str(simple_label(k))
+        for k in val_vids.keys()
```
Comment (dreamerlin) [inline tools/data/activitynet/generate_rawframes_filelist.py:66]: `for k in val_vids:`

File: tools/data/activitynet/generate_rawframes_filelist.py [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+    ]
+
+    with open(osp.join(data_file, 'anet_train_video.txt'), 'w') as fout:
+        fout.write('\n'.join(train_lines))
+    with open(osp.join(data_file, 'anet_val_video.txt'), 'w') as fout:
+        fout.write('\n'.join(val_lines))
+
+    def clip_list(k, anno, vidanno):
+        n_second = anno['duration_second']
```
Comment (dreamerlin) [inline tools/data/activitynet/generate_rawframes_filelist.py:75]: `n_second` -> `num_seconds`

File: tools/data/activitynet/generate_rawframes_filelist.py [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+        n_frame = vidanno['numframe']
```
Comment (dreamerlin) [inline tools/data/activitynet/generate_rawframes_filelist.py:76]: `n_frame` -> `num_frames`

File: tools/data/activitynet/generate_rawframes_filelist.py [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+        fps = n_frame / n_second
+        segs = anno['annotations']
+        lines = []
+        for seg in segs:
+            segment = seg['segment']
+            lb = seg['label']
```
Comment (dreamerlin) [inline tools/data/activitynet/generate_rawframes_filelist.py:82]: `lb` -> `label`

File: tools/data/activitynet/generate_rawframes_filelist.py [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+            lb = anet_labels.index(lb)
+            start, end = int(segment[0] * fps), int(segment[1] * fps)
+            if end > n_frame - 1:
+                end = n_frame - 1
+            newline = f'{k} {start} {end - start + 1} {lb}'
+            lines.append(newline)
+        return lines
+
+    train_clips, val_clips = [], []
+    for k in train_vids.keys():
```
Comment (dreamerlin) [inline tools/data/activitynet/generate_rawframes_filelist.py:92]: `for k in train_vids:`

File: tools/data/activitynet/generate_rawframes_filelist.py [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+        train_clips.extend(clip_list(k, anet_annotations[k], train_vids[k]))
+    for k in val_vids.keys():
```
Comment (dreamerlin) [inline tools/data/activitynet/generate_rawframes_filelist.py:94]: `for k in val_vids:`

File: tools/data/activitynet/download.py (hunk 6 above)
Comment (kennymckormick) [reply tools/data/activitynet/download.py:2]: It is MIT license. How to properly add it to the head of code? I browsed some repos and didn't find the answer.

File: tools/data/activitynet/download.py (hunk 7 above)
Comment (kennymckormick) [reply tools/data/activitynet/download.py:105]: some problems also exist in Kinetics datatools, I also fixed them

File: tools/data/activitynet/generate_rawframes_filelist.py (hunk 10 above)
Comment (kennymckormick) [reply tools/data/activitynet/generate_rawframes_filelist.py:44]: bug in original code, fix

File: tools/data/activitynet/generate_rawframes_filelist.py (hunk 12 above)
Comment (kennymckormick) [reply tools/data/activitynet/generate_rawframes_filelist.py:55]: Fixed, and where can I find some reference about how to properly name those variables, I think some abbreviations should be OK?

File: tools/data/activitynet/download.py (hunk 6 above)
Comment (innerlee) [reply tools/data/activitynet/download.py:2]: add a line of comment like
```
The code is licensed under the MIT licence.
```

**Review by innerlee — APPROVED** [review]

**Pull request merged**